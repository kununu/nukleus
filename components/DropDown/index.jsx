import React, {PropTypes, Component} from 'react';

import styles from './index.scss';

import {isBrowser} from '../../utils/executionEnvironment';
import {
  clearfix
} from '../index.scss';

export default class Dropdown extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.shape({
      active: PropTypes.boolean,
      icon: PropTypes.element,
      link: PropTypes.element.isRequired,
      value: PropTypes.string.isRequired
    })).isRequired,
    position: PropTypes.oneOf(['top', 'bottom']),
    shade: PropTypes.oneOf(['light', 'dark'])
  };

  static defaultProps = {
    position: 'top',
    shade: 'light'
  }

  state = {
    isOpen: false
  }

  componentWillMount () {
    if (!isBrowser) return;
    document.addEventListener('click', this.onClickDocument, false);
    document.addEventListener('touchend', this.onClickDocument, false);
  }

  componentWillUnmount () {
    if (!isBrowser) return;
    document.removeEventListener('click', this.onClickDocument, false);
    document.removeEventListener('touchend', this.onClickDocument, false);
  }

  onButtonClick = () => {
    this.setState({isOpen: !this.state.isOpen});
  }

  onClickDocument = e => {
    if (this.isButtonElement(e)) return;
    this.setState({isOpen: false});
  }

  getActiveItem () {
    return this.props.items.filter(item => item.active)[0];
  }

  getItem = item => (
    <span className={clearfix}>
      <span className={styles.pullLeft}>
        {item.value}
      </span>
      {item.icon ?
        <span className={styles.pullRight}>
          {item.icon}
        </span>
        : ''}
    </span>
    )

  isButtonElement (e) {
    return this.node.contains(e.target);
  }

  render () {
    const {
      items,
      position,
      shade
    } = this.props;
    const activeItem = this.getActiveItem();

    return (
      <div className={`${styles.container} ${styles[position]}`}>
        <button
          ref={node => this.node = node}
          className={`${styles.selection} ${clearfix} ${styles[shade]}`}
          onClick={this.onButtonClick}>
          {activeItem.value}
          {activeItem.icon ?
            <span className={styles.pullRight}>
              {activeItem.icon}
            </span>
          : ''}
        </button>
        <ul className={`${styles.menu} ${this.state.isOpen ? styles.open : ''}`}>
          {items.map(item =>
            <li // eslint-disable-line
              key={item.value}
              className={`${styles.item} ${clearfix}`}>
              {React.cloneElement(item.link, [], this.getItem(item))}
            </li>
          )}
        </ul>
      </div>
    );
  }
}
