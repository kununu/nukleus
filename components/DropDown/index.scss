@import '~kununu-theme/scss/base/variables';

.container {
  position: relative;
  display: inline-block;
}

.container.top {
  .menu {
    top: auto;
    bottom: 100%;
    border-radius: 4px 4px 0 0;
  }
}

.container.bottom {
  .menu {
    border-radius: 0 0 4px 4px;
  }
}

.selection {
  border: 0;
  background-color: transparent;
  padding: 12px 0 12px 12px;
  font-weight: $font-weight-semibold;
  line-height: 20px;
  min-width: 130px;
  display: flex;
  align-items: center;
  justify-content: flex-end;

  span {
    padding-left: 10px;
  }

  &.light {
    color: $white;
  }

  &.dark {
    color: $text-color;
  }

  &,
  &:focus {
    outline: none;
  }
}

.menu {
  background-color: $white;
  display: none;
  min-width: 160px;
  margin: 0;
  padding: 0;
  position: absolute;
  right: 0;
  z-index: $zindex-popover;

  li {
    list-style-type: none;
  }

  &.open {
    display: block;
  }
}

.item {
  cursor: pointer;
  font-size: $font-size-small;
  color: $gray-dark;

  &:hover {
    background-color: $gray-base-03;
  }

  a {
    display: block;
    padding: 10px 20px;

    &:first-child {
      padding-top: 20px;
    }

    &:hover {
      text-decoration: none;
      color: inherit;
    }
  }
}

.pullLeft {
  float: left;
}

.pullRight {
  float: right;
  text-align: right;
}
