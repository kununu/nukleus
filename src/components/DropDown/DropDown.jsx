import React, {Component} from 'react';
import PropTypes from 'prop-types';

import {isBrowser} from 'utils/executionEnvironment';

import styles from './index.scss';

export default class DropDown extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.element),
      PropTypes.element
    ]).isRequired,
    direction: PropTypes.oneOf(['up', 'down']),
    pullRight: PropTypes.bool,
    showOnHover: PropTypes.bool,
    title: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element
    ]).isRequired,
    type: PropTypes.oneOf(['default', 'countrySwitcher'])
  }

  static defaultProps = {
    direction: 'down',
    pullRight: false,
    showOnHover: true,
    type: 'default'
  }

  state = {isOpen: false}

  componentDidMount () {
    if (!isBrowser) return;

    document.addEventListener('click', this.handleClickOutside, false);
    document.addEventListener('touchend', this.handleClickOutside, false);
  }

  componentWillUnmount () {
    if (!isBrowser) return;

    document.removeEventListener('click', this.handleClickOutside, false);
    document.removeEventListener('touchend', this.handleClickOutside, false);
  }

  onMouseEnter = () => {
    const {showOnHover} = this.props;

    if (!showOnHover) return;

    this.open();
  }

  onMouseLeave = () => {
    const {showOnHover} = this.props;

    if (!showOnHover) return;

    this.close();
  }

  onClick = () => this.setState({isOpen: !this.state.isOpen});

  handleClickOutside = e => {
    if (this.node.contains(e.target)) return;

    setTimeout(() => {
      this.close();
    }, 100);
  }

  open = () => this.setState({isOpen: true});

  close = () => this.setState({isOpen: false});

  render () {
    const {
      children,
      direction,
      pullRight,
      title,
      type
    } = this.props;
    const {isOpen} = this.state;

    return (
      <div className={`${styles.container} ${styles[direction]} ${pullRight ? styles.pullRight : ''} ${type === 'countrySwitcher' ? styles.countrySwitcher : ''}`}>
        <button
          type="button"
          id="dropdown"
          className={styles.toggleButton}
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
          onClick={this.onClick}
          ref={node => this.node = node}
          aria-haspopup="listbox"
          aria-expanded={isOpen}>
          {title}
        </button>

        {isOpen &&
          <ul
            className={styles.itemsList}
            onMouseEnter={this.onMouseEnter}
            onMouseLeave={this.onMouseLeave}
            aria-labelledby="dropdown"
            role="listbox"
            tabIndex={0}>
            {children}
          </ul>
        }
      </div>
    );
  }
}
