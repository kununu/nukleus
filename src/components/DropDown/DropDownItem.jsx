import React from 'react';
import PropTypes from 'prop-types';

import styles from './index.scss';

export default function DropDownItem ({children, icon, isActive}) {
  return (
    <div className={`${styles.item} ${isActive && styles.active} ${icon && styles.hasIcon}`}>
      {children}

      {icon &&
        <span className={styles.icon}>
          {icon}
        </span>
      }
    </div>
  );
}

DropDownItem.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired,
  icon: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.bool,
  ]),
  isActive: PropTypes.bool
};

DropDownItem.defaultProps = {
  icon: false,
  isActive: false
};
