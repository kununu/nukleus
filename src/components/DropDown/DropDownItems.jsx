import React from 'react';
import PropTypes from 'prop-types';

import styles from './index.scss';

export default function DropDownItems ({children}) {
  return (
    <div className={styles.items}>
      {children}
    </div>
  );
}

DropDownItems.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired
};
