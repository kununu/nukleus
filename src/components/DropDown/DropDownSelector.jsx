import React from 'react';
import PropTypes from 'prop-types';

import styles from './index.scss';

export default function DropDownSelector ({children, isActive}) {
  return (
    <div className={`${styles.selectedItem} ${isActive && styles.activeHeader}`}>
      {children}
    </div>
  );
}

DropDownSelector.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element
  ]).isRequired,
  isActive: PropTypes.bool
};

DropDownSelector.defaultProps = {isActive: false};
