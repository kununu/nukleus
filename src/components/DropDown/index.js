export {default as DropDown} from './DropDown';
export {default as DropDownItem} from './DropDownItem';
