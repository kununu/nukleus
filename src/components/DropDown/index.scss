@import '~@kununu/kununu-theme-v2/scss/base/variables';

// .container {
//   position: relative;
//   display: inline-block;
// }
//
// .container.top {
//   .menu {
//     top: auto;
//     bottom: 100%;
//     border-radius: 4px 4px 0 0;
//   }
// }
//
// .container.bottom {
//   .menu {
//     border-radius: 0 0 4px 4px;
//   }
// }
//
// .selection {
//   border: 0;
//   background-color: transparent;
//   padding: 12px 0 12px 12px;
//   font-weight: $font-weight-semibold;
//   line-height: 20px;
//   min-width: 130px;
//   display: flex;
//   align-items: center;
//   justify-content: flex-end;
//
//   span {
//     padding-left: 10px;
//   }
//
//   &.light {
//     color: $white;
//   }
//
//   &.dark {
//     color: $text-color;
//   }
//
//   &,
//   &:focus {
//     outline: none;
//   }
// }
//
// .menu {
//   background-color: $white;
//   display: none;
//   min-width: 160px;
//   margin: 0;
//   padding: 0;
//   position: absolute;
//   right: 0;
//   z-index: 1060;
//   overflow: hidden;
//
//   li {
//     list-style-type: none;
//   }
//
//   &.open {
//     display: block;
//   }
// }
//
// .item {
//   cursor: pointer;
//   font-size: $font-size-small;
//   color: $text-color;
//
//   &:hover {
//     background-color: $gray-base-03;
//   }
//
//   &:first-child a {
//     padding-top: 20px;
//   }
//
//   a {
//     display: block;
//     padding: 10px 20px;
//
//     &,
//     &:hover {
//       text-decoration: none;
//       color: inherit;
//     }
//   }
// }

@import '~@kununu/kununu-theme-v2/scss/base/variables';


.container {
  position: relative;
  display: inline-block;

  &.top {
    .items {
      top: auto;
      bottom: 40px;
      left: -50px;
      border-radius: $border-radius-base 0 0;
    }
  }

  &.bottom {
    .items {
      border-radius: 0 0 $border-radius-base;
    }
  }

  &.light {
    .selectedItem a {
      color: $white;
    }
  }

  &.dark {
    .selectedItem a {
      color: $text-color;
    }
  }
}

.selectedItem {
  .item {
    a {
      padding: 10px 30px;
      font-weight: $font-weight-semibold;
      text-align: right;

      &:hover {
        background-color: none;
      }
    }

    &:hover {
      background-color: transparent;
    }
  }
}

.items {
  position: absolute;
  top: 40px;
  right: -10px;
  min-width: 160px;
  background-color: $white;
  box-shadow: 0 7px 15px 0 rgba(0, 0, 0, .1);
  z-index: 1060;

  .item {
    &.hasIcon a {
      padding-right: 40px;
    }

    .icon {
      right: 20px;
    }
  }
}

.item {
  position: relative;
  font-size: $font-size-small;
  cursor: pointer;

  .icon {
    position: absolute;
    top: 50%;
    right: 0;
    margin-top: -9px;
  }

  a {
    display: block;
    padding: 10px 20px;
    color: $text-color;
    text-align: left;

    &:hover,
    &:active,
    &:focus,
    &:visited {
      text-decoration: none;
    }
  }

  &.active,
  &:hover {
    background-color: $gray-base-03;
  }
}

.activeHeader a {
  color: $teal-dark;

  &:focus {
    text-decoration: none;
  }
}
