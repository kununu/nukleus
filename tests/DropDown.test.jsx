import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import DropDown from 'DropDown'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies


const items = [
  {
    icon: <img
      alt="Austrian flag"
      src="austrianflag.jpg" />,
    link: <a href="/at" alt="hi">Austria</a>,
    value: 'Austria'
  },
  {
    icon: <img alt="German flag" src="germanflag.jpg" />,
    link: <a href="/de">German</a>,
    value: 'German'
  }
];

test('Renders DropDown without crashing', () => {
  const component = renderer.create(
    <DropDown
      pathname="/at"
      items={items} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders correct default selection', () => {
  const component = renderer.create(
    <DropDown
      pathname="/at"
      items={[
        {
          icon: <img alt="Austrian flag" src="austrianflag.jpg" />,
          link: <a href="/at" alt="hi">Austria</a>,
          value: 'Austria'
        },
        {
          icon: <img alt="German flag" src="germanflag.jpg" />,
          link: <a href="/de">German</a>,
          value: 'German'
        }
      ]} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders menu when button is clicked', () => {
  const component = mount(
    <DropDown
      pathname="/at"
      items={items} />
  );
  component.find('button').simulate('click');
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('Closes menu when document is clicked', () => {
  const component = mount(
    <DropDown
      pathname="/at"
      items={items} />
  );
  component.find('button').simulate('click');
  component.find('button').simulate('click');
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

test('Renders correctly when no icon is provided', () => {
  const component = renderer.create(
    <DropDown
      pathname="/at"
      items={[
        {
          active: false,
          link: <a href="/at" alt="hi">Austria</a>,
          value: 'Austria'
        },
        {
          active: false,
          link: <a href="/de">German</a>,
          value: 'German'
        }
      ]} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Updates selection on item click', () => {
  const component = mount(
    <DropDown
      pathname="/at"
      items={items} />
  );
  component.last('li').simulate('click');
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
