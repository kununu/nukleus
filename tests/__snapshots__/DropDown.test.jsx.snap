exports[`test Closes menu when document is clicked 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "active": false,
        "icon": <img
          alt="Austrian flag"
          src="austrianflag.jpg" />,
        "link": <a
          alt="hi"
          href="/at">
          Austria
    </a>,
        "value": "Austria",
      },
      Object {
        "active": true,
        "icon": <img
          alt="German flag"
          src="germanflag.jpg" />,
        "link": <a
          href="/de">
          German
    </a>,
        "value": "German",
      },
    ]
  }
  position="top"
  shade="light">
  <div
    className="container top">
    <button
      className="selection clearfix light"
      onClick={[Function]}>
      German
      <span
        className="pullRight">
        <img
          alt="German flag"
          src="germanflag.jpg" />
      </span>
    </button>
    <ul
      className="menu ">
      <li
        className="item clearfix">
        <a
          alt="hi"
          href="/at">
          <span
            className="clearfix">
            <span
              className="pullLeft">
              Austria
            </span>
            <span
              className="pullRight">
              <img
                alt="Austrian flag"
                src="austrianflag.jpg" />
            </span>
          </span>
        </a>
      </li>
      <li
        className="item clearfix">
        <a
          href="/de">
          <span
            className="clearfix">
            <span
              className="pullLeft">
              German
            </span>
            <span
              className="pullRight">
              <img
                alt="German flag"
                src="germanflag.jpg" />
            </span>
          </span>
        </a>
      </li>
    </ul>
  </div>
</Dropdown>
`;

exports[`test Renders DropDown without crashing 1`] = `
<div
  className="container top">
  <button
    className="selection clearfix light"
    onClick={[Function]}>
    German
    <span
      className="pullRight">
      <img
        alt="German flag"
        src="germanflag.jpg" />
    </span>
  </button>
  <ul
    className="menu ">
    <li
      className="item clearfix">
      <a
        alt="hi"
        href="/at">
        <span
          className="clearfix">
          <span
            className="pullLeft">
            Austria
          </span>
          <span
            className="pullRight">
            <img
              alt="Austrian flag"
              src="austrianflag.jpg" />
          </span>
        </span>
      </a>
    </li>
    <li
      className="item clearfix">
      <a
        href="/de">
        <span
          className="clearfix">
          <span
            className="pullLeft">
            German
          </span>
          <span
            className="pullRight">
            <img
              alt="German flag"
              src="germanflag.jpg" />
          </span>
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`test Renders correct default selection 1`] = `
<div
  className="container top">
  <button
    className="selection clearfix light"
    onClick={[Function]}>
    Austria
    <span
      className="pullRight">
      <img
        alt="Austrian flag"
        src="austrianflag.jpg" />
    </span>
  </button>
  <ul
    className="menu ">
    <li
      className="item clearfix">
      <a
        alt="hi"
        href="/at">
        <span
          className="clearfix">
          <span
            className="pullLeft">
            Austria
          </span>
          <span
            className="pullRight">
            <img
              alt="Austrian flag"
              src="austrianflag.jpg" />
          </span>
        </span>
      </a>
    </li>
    <li
      className="item clearfix">
      <a
        href="/de">
        <span
          className="clearfix">
          <span
            className="pullLeft">
            German
          </span>
          <span
            className="pullRight">
            <img
              alt="German flag"
              src="germanflag.jpg" />
          </span>
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`test Renders correctly when no icon is provided 1`] = `
<div
  className="container top">
  <button
    className="selection clearfix light"
    onClick={[Function]}>
    Austria
    
  </button>
  <ul
    className="menu ">
    <li
      className="item clearfix">
      <a
        alt="hi"
        href="/at">
        <span
          className="clearfix">
          <span
            className="pullLeft">
            Austria
          </span>
          
        </span>
      </a>
    </li>
    <li
      className="item clearfix">
      <a
        href="/de">
        <span
          className="clearfix">
          <span
            className="pullLeft">
            German
          </span>
          
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`test Renders menu when button is clicked 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "active": false,
        "icon": <img
          alt="Austrian flag"
          src="austrianflag.jpg" />,
        "link": <a
          alt="hi"
          href="/at">
          Austria
    </a>,
        "value": "Austria",
      },
      Object {
        "active": true,
        "icon": <img
          alt="German flag"
          src="germanflag.jpg" />,
        "link": <a
          href="/de">
          German
    </a>,
        "value": "German",
      },
    ]
  }
  position="top"
  shade="light">
  <div
    className="container top">
    <button
      className="selection clearfix light"
      onClick={[Function]}>
      German
      <span
        className="pullRight">
        <img
          alt="German flag"
          src="germanflag.jpg" />
      </span>
    </button>
    <ul
      className="menu open">
      <li
        className="item clearfix">
        <a
          alt="hi"
          href="/at">
          <span
            className="clearfix">
            <span
              className="pullLeft">
              Austria
            </span>
            <span
              className="pullRight">
              <img
                alt="Austrian flag"
                src="austrianflag.jpg" />
            </span>
          </span>
        </a>
      </li>
      <li
        className="item clearfix">
        <a
          href="/de">
          <span
            className="clearfix">
            <span
              className="pullLeft">
              German
            </span>
            <span
              className="pullRight">
              <img
                alt="German flag"
                src="germanflag.jpg" />
            </span>
          </span>
        </a>
      </li>
    </ul>
  </div>
</Dropdown>
`;

exports[`test Updates selection on item click 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "active": false,
        "icon": <img
          alt="Austrian flag"
          src="austrianflag.jpg" />,
        "link": <a
          alt="hi"
          href="/at">
          Austria
    </a>,
        "value": "Austria",
      },
      Object {
        "active": true,
        "icon": <img
          alt="German flag"
          src="germanflag.jpg" />,
        "link": <a
          href="/de">
          German
    </a>,
        "value": "German",
      },
    ]
  }
  position="top"
  shade="light">
  <div
    className="container top">
    <button
      className="selection clearfix light"
      onClick={[Function]}>
      German
      <span
        className="pullRight">
        <img
          alt="German flag"
          src="germanflag.jpg" />
      </span>
    </button>
    <ul
      className="menu ">
      <li
        className="item clearfix">
        <a
          alt="hi"
          href="/at">
          <span
            className="clearfix">
            <span
              className="pullLeft">
              Austria
            </span>
            <span
              className="pullRight">
              <img
                alt="Austrian flag"
                src="austrianflag.jpg" />
            </span>
          </span>
        </a>
      </li>
      <li
        className="item clearfix">
        <a
          href="/de">
          <span
            className="clearfix">
            <span
              className="pullLeft">
              German
            </span>
            <span
              className="pullRight">
              <img
                alt="German flag"
                src="germanflag.jpg" />
            </span>
          </span>
        </a>
      </li>
    </ul>
  </div>
</Dropdown>
`;
